### PR TITLE
mobile: Add an optional idempotency parameter in sendHeaders

### PR DESCRIPTION
**Pull request by fredyw**
This PR adds an optional parameter to indicate whether a request is an idempotent in the `sendHeaders`. This can be useful to automatically add a retry policy, such as `http3-post-connect-failure` when the `idempotent` flag is set. This makes the Envoy Mobile API similar to Cronet API. The Cronvoy implementation has also been updated to respect the `idempotency` flag.

Risk Level: low
Testing: unit test
Docs Changes: n/a
Release Notes: n/a
Platform Specific Features: mobile

### CODE REVIEW

File: mobile/library/java/org/chromium/net/impl/CronvoyUrlRequest.java [hunk 1]
```diff
@@ -210,6 +211,7 @@ public final class CronvoyUrlRequest extends CronvoyUrlRequestBase {
     mCurrentUrl = url;
     mUserAgent = userAgent;
     mRequestAnnotations = connectionAnnotations;
+    mIempotent = idempotent;
```
Comment (alyssawilk) [inline mobile/library/java/org/chromium/net/impl/CronvoyUrlRequest.java:214]: mIdempotent?

File: mobile/library/cc/stream.h [hunk 2]
```diff
@@ -21,8 +21,10 @@ class Stream {
    *
    * @param headers the headers to send.
    * @param end_stream indicates whether to close the stream locally after sending this frame.
+   * @param idempotent indicates that the request is idempotent. When idempotent is set to true
+   *                   Envoy Mobile may perform retry on failures. By default, it is set to false.
```
Comment (alyssawilk) [inline mobile/library/cc/stream.h:25]: can we qualify
HTTP/3 post-handshake failures
?
We already retry on failures automatically for safe requests for example

File: mobile/library/java/io/envoyproxy/envoymobile/engine/EnvoyHTTPStream.java [hunk 3]
```diff
@@ -38,11 +38,14 @@ public EnvoyHTTPStream(long engineHandle, long streamHandle, EnvoyHTTPCallbacks
    * Send headers over an open HTTP streamHandle. This method can be invoked once
    * and needs to be called before send_data.
    *
-   * @param headers,   the headers to send.
-   * @param endStream, supplies whether this is headers only.
+   * @param headers    the headers to send.
+   * @param endStream  supplies whether this is headers only.
+   * @param idempotent indicates that the request is idempotent. When idempotent is set to true
+   *                   Envoy Mobile may perform retry on failures. By default, it is set to false.
```
Comment (alyssawilk) [inline mobile/library/java/io/envoyproxy/envoymobile/engine/EnvoyHTTPStream.java:44]: ditto on retry comments, here and below

File: mobile/library/java/org/chromium/net/impl/CronvoyUrlRequestContext.java [hunk 4]
```diff
@@ -132,7 +132,8 @@ void setTaskToExecuteWhenInitializationIsCompleted(Runnable runnable) {
       checkHaveAdapter();
       return new CronvoyUrlRequest(this, url, callback, executor, mUserAgent, allowDirectExecutor,
                                    requestAnnotations, trafficStatsTagSet, trafficStatsTag,
-                                   trafficStatsUidSet, trafficStatsUid, requestFinishedListener);
+                                   trafficStatsUidSet, trafficStatsUid, requestFinishedListener,
+                                   idempotency == 1);
```
Comment (alyssawilk) [inline mobile/library/java/org/chromium/net/impl/CronvoyUrlRequestContext.java:136]: can you use or comment-link to ExperimentalUrlRequest.java IDEMPOTENT definition?

**Review by alyssawilk — COMMENTED** [review]
nice!

File: mobile/library/java/org/chromium/net/impl/CronvoyUrlRequest.java (hunk 1 above)
Comment (fredyw) [reply mobile/library/java/org/chromium/net/impl/CronvoyUrlRequest.java]: Done.

File: mobile/library/cc/stream.h (hunk 2 above)
Comment (fredyw) [reply mobile/library/cc/stream.h]: Done.

File: mobile/library/java/io/envoyproxy/envoymobile/engine/EnvoyHTTPStream.java (hunk 3 above)
Comment (fredyw) [reply mobile/library/java/io/envoyproxy/envoymobile/engine/EnvoyHTTPStream.java]: Done.

File: mobile/library/java/org/chromium/net/impl/CronvoyUrlRequestContext.java (hunk 4 above)
Comment (fredyw) [reply mobile/library/java/org/chromium/net/impl/CronvoyUrlRequestContext.java]: Done. I updated the code to use the constant from Cronet API.

**Review by abeyad — APPROVED** [review]

**Review by alyssawilk — APPROVED** [review]

**Pull request merged**